homeassistant/helpers/entity.py

```diff
@@ -68,14 +43,26 @@ def async_generate_entity_id(
     hass: Optional[HomeAssistant] = None,
 ) -> str:
     """Generate a unique entity ID based on given entity IDs or used IDs."""
-    if current_ids is None:
-        if hass is None:
-            raise ValueError("Missing required parameter currentids or hass")
 
-        current_ids = hass.states.async_entity_ids()
     name = (name or DEVICE_DEFAULT_NAME).lower()
+    preferred_string = entity_id_format.format(slugify(name))
+
+    if current_ids is not None:
+        return ensure_unique_string(preferred_string, current_ids)
+
+    if hass is None:
+        raise ValueError("Missing required parameter currentids or hass")
+
+    test_string = preferred_string
+    tries = 1
+    while hass.states.get(test_string):
+        tries += 1
+        test_string = f"{preferred_string}_{tries}"
+
+    return test_string
+
 
-    return ensure_unique_string(entity_id_format.format(slugify(name)), current_ids)
+generate_entity_id = async_generate_entity_id  # pylint: disable=invalid-name
```

Comment:
`async_generate_entity_id` is decorated with `callback`. If we set `generate_entity_id` to the same name, it will also be get that decoration and that will be misleading as it doesn't have to run within the event loop.

Reply:
Will make it a wrapper